### PR TITLE
sridharb1_contrib_msvc_0.27

### PR DESCRIPTION
**Pull request by clanmills**

### CODE REVIEW

**Comment by sridharb1** [comment]
Thanks Robin. Will look into this soon, definitely by Monday.


Sent from Yahoo Mail for iPhone


On Saturday, March 28, 2020, 21:46, Robin Mills <notifications@github.com> wrote:




@clanmills requested your review on: #1120 sridharb1_contrib_msvc_0.27.

**Review by sridharb1 — APPROVED** [review]
Looks good. Please merge, that will allow me to test locally.

**Comment by sridharb1** [comment]
Please also apply this small patch as well.


[license.patch.txt](https://github.com/Exiv2/exiv2/files/4398429/license.patch.txt)

File: contrib/msvc/LICENSE
Comment (D4N) [inline contrib/msvc/LICENSE]:
```suggestion
Copyright (c) 2020 Sridhar Boovaraghavan
```

**Review by D4N — COMMENTED** [review]
Please consider adding Sridhar's github handle as the codeowner for `contrib/msvc` into a `CODEOWNERS` file (see https://help.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners).

File: contrib/msvc/LICENSE
Comment (clanmills) [reply contrib/msvc/LICENSE]: I've modified this to be:  
```
Copyright (c) 2020 Sridhar Boovaraghavan  GitHub Handle: sridharb1
```
@sridharb1 Please defend your change, provide alternative or approve my version.

**Comment by clanmills** [comment]
I agree with you, @D4N about CODEOWNERS and I've opened a new issue.  I will not delay merging this PR to work on #1121.

**Comment by clanmills** [comment]
@sridharb1 Please test this on your machine.  I haven't tested this yet, however I will before RC1.

It's your responsibility to maintain this contribution.  There are very few contribs and we only have resources to inspect them to ensure they have good intent (without malware or virus).  You have write access to Exiv2 GitHub and can submit PRs to change `contrib/msvc`

**Review by sridharb1 — APPROVED** [review]

**Comment by sridharb1** [comment]
Is it possible for you to link this to my exiv2-x86_x64 project? If so, it would be easy for me to maintain it as I would make changes there and I can "pull" them into this area (hopefully easily).

Let me warn you that testing this will take a fair amount of work, only because it will involve recreating the dependency tree that I detail in the README.md. However, it should not be "hard", which is the point of me creating these projects for others to be able to easily compile on Windows.

**Comment by sridharb1** [comment]
There are three changes required. I have attached those patch files because I cannot push to this folder in 0.27-maintenance. 


[1.patch.txt](https://github.com/Exiv2/exiv2/files/4399190/1.patch.txt)
[2.patch.txt](https://github.com/Exiv2/exiv2/files/4399191/2.patch.txt)
[3.patch.txt](https://github.com/Exiv2/exiv2/files/4399192/3.patch.txt)

**Comment by sridharb1** [comment]
Just for clarity, what rights to these files do I have now? Am I able to push changes to these? If not, it doesn't make sense to me if I have to submit patches like before to a different maintainer (who is not interested in these files anyway), while I am supposed to be the "official" maintainer.

**Comment by clanmills** [comment]
@sridharb1 I've submitted those changes using a new PR.  I don't think it's possible to reopen a PR after it's been merged and deleted.  I tried and couldn't get it to work, so I opened a new PR. #1128  There's something odd about 1.patch.txt which wants to create a new file `mrwthumb.vcxproj` which already exists.  I've replaced the old file.

You have write access to the Exiv2 GitHub repos and that should be sufficient access for your purposes.  I am an "administrator" which gives me write access and enables me to change settings, add/remove users and other features.

I don't understand git, however I'm friendly and helpful.  I'll help you to feel comfortable with git.  Things will be fine - you'll see.  I use a mix of command-line and SmartGit.  I recommend you use SmartGit.  There is a free license for open-source projects.  Almost everything you need to know about git is on my little card.
GitIdiotCard


It's pointless to criticise or complain about git.  It's a religion in the open source community.  You will incur the hatred of the community if you say anything bad about git.  And much the same (in reverse) about Visual Studio which is loathed by the community.  Few in the community have bothered to understand it and hate it for religious reasons.

**Comment by clanmills** [comment]
@sridharb1 Can I say a few words in praise of CMake.  I agree with you that "native" Visual Studio sln/vcxproj files are much superior to the files generated by CMake and that's why I supported them for Exiv2 for about 10 years.  However, they are a maintenance head-ache for the project.  Until conan arrived, using CMake with Visual Studio was effectively 100% broken.  I put 200 hours into the development of a script `cmake_build.bat` to make it possible to build Exiv2 and dependencies with CMake.

When @piponazo joined the project, he advocated for conan, rewrote our CMakeLists.txt files and added the cmake/ directory.  I agreed to remove the msvc sln/vcxproj files and the unix autotools.  The build has been more stable and less time consuming.  Luis is right.  This is better from the point of view of managing the Exiv2 project.

I do my day-to-day development and debugging on macOS.  I had a free open-source license for CLion which is a cross platform IDE with very good CMake support.  The license expired after I retired and I didn't request a renewal.  Today I'm using Xcode which is free.  CMake generates Xcode project files that work well and enable me to debug the code.  I'm happy with CMake.

**Comment by sridharb1** [comment]
> 
> 
> @sridharb1 I've submitted those changes using a new PR. I don't think it's possible to reopen a PR after it's been merged and deleted. I tried and couldn't get it to work, so I opened a new PR. #1128 There's something odd about 1.patch.txt which wants to create a new file `mrwthumb.vcxproj` which already exists. I've replaced the old file.
> 
Don't know and can't explain. All I did was to pull down the 0.27-maintenance branch after you had merged it and compared my build directory with the contents of contrib/msvc and found that this was missing.

Regarding access, there is a problem that would make this arrangement not sustainable. If I am the maintainer of this directory for the 0.27-maintenance branch, then I need write access to this folder. If not, we are back to square one of me contributing changes, inclusion of which depends on the whims of someone else. I have had that discussion of why I think this is useful and am not inclined to pursue it again with people who disagree.

So, this needs to be clear going forward. If I am the maintainer of this (and I take this responsibility for the 0.27 branch because I haven't invested time into supporting changes that are in 0.28+), I need to be able to push into this hierarchy/folder (don't care about this permission in the other folders, the usual process is fine). If I can't, then, no deal.

**Comment by sridharb1** [comment]
> 
> 
> @sridharb1 Can I say a few words in praise of CMake. I agree with you that "native" Visual Studio sln/vcxproj files are much superior to the files generated by CMake and that's why I supported them for Exiv2 for about 10 years. However, they are a maintenance head-ache for the project. Until conan arrived, using CMake with Visual Studio was effectively 100% broken. I put 200 hours into the development of a script `cmakeBuild.cmd` to make it possible to build Exiv2 and dependencies with CMake.
> 
CMake does a good job (I won't give it full marks yet - I do use it in Linux environments, where I would like it to be better at certain things). I understand how it makes the maintainers' life easier. I also understand the decision to use it as the default build system. I always knew that its generated files for VS have its limitations.

What got my goat is the fact that a perfectly good VS solution (not the cmakebuild.cmd, which was an ill-fated attempt to satisfy both worlds and met its preordained fate) was thrown out because of this "xenophobia". It was in the contrib folder. It was doing no harm to anything. People like me would have used it every now and then, found that it was missing one or another file, added it and possibly update you all on the changes.

However, this purity virus creates problems for people and as long as those problems are not yours, one doesn't care. That is not acceptable.

Anyway, it is not worth chewing the cud over this and I have moved on. My current "issues" have been detailed in the previous message.

**Comment by D4N** [comment]
Sridhar Boovaraghavan <notifications@github.com> writes:

>> 
>> 
>> @sridharb1 I've submitted those changes using a new PR. I don't think it's possible to reopen a PR after it's been merged and deleted. I tried and couldn't get it to work, so I opened a new PR. #1128 There's something odd about 1.patch.txt which wants to create a new file `mrwthumb.vcxproj` which already exists. I've replaced the old file.
>> 
> Don't know and can't explain. All I did was to pull down the 0.27-maintenance branch after you had merged it and compared my build directory with the contents of contrib/msvc and found that this was missing.
>
> Regarding access, there is a problem that would make this arrangement not sustainable. If I am the maintainer of this directory for the 0.27-maintenance branch, then I need write access to this folder. If not, we are back to square one of me contributing changes, inclusion of which depends on the whims of someone else. I have had that discussion of why I think this is useful and am not inclined to pursue it again with people who disagree.
>
> So, this needs to be clear going forward. If I am the maintainer of this (and I take this responsibility for the 0.27 branch because I haven't invested time into supporting changes that are in 0.28+), I need to be able to push into this hierarchy/folder (don't care about this permission in the other folders, the usual process is fine). If I can't, then, no deal.
>

We have established the pull request workflow to ensure that only
non-broken commits get into the master or maintenance branches. You
already got a pass on the "only tested" part (as your project files are
not tested on the CI), so please consider sticking to the pull request
workflow as unfortunately github does not provide per folder ACLs. But a
human reviewer can easily see and ACK your PR.

**Comment by sridharb1** [comment]
> 
> Sridhar Boovaraghavan <notifications@github.com> writes:
> > > > @sridharb1 I've submitted those changes using a new PR. I don't think it's possible to reopen a PR after it's been merged and deleted. I tried and couldn't get it to work, so I opened a new PR. #1128 There's something odd about 1.patch.txt which wants to create a new file `mrwthumb.vcxproj` which already exists. I've replaced the old file. > Don't know and can't explain. All I did was to pull down the 0.27-maintenance branch after you had merged it and compared my build directory with the contents of contrib/msvc and found that this was missing. Regarding access, there is a problem that would make this arrangement not sustainable. If I am the maintainer of this directory for the 0.27-maintenance branch, then I need write access to this folder. If not, we are back to square one of me contributing changes, inclusion of which depends on the whims of someone else. I have had that discussion of why I think this is useful and am not inclined to pursue it again with people who disagree. So, this needs to be clear going forward. If I am the maintainer of this (and I take this responsibility for the 0.27 branch because I haven't invested time into supporting changes that are in 0.28+), I need to be able to push into this hierarchy/folder (don't care about this permission in the other folders, the usual process is fine). If I can't, then, no deal.
> We have established the pull request workflow to ensure that only non-broken commits get into the master or maintenance branches. You already got a pass on the "only tested" part (as your project files are not tested on the CI), so please consider sticking to the pull request workflow as unfortunately github does not provide per folder ACLs. But a human reviewer can easily see and ACK your PR.

That's fair. However, let me propose the following:

1. link contrib/msvc to my github repository. That is, fork sridharb1/exiv2-x86_x64 to contrib/msvc in the exiv2 repository as a submodule.
2. This way, my changes in my repository could be pulled (I could initiate a pull request or release admins can, when they are getting close to a RC1).
3. This preserves the control that current administrators have over the exiv2 repository while delinking it from rights that I might need to properly maintain it because I would maintain my own repository.

Is this possible?

**Comment by D4N** [comment]
Sridhar Boovaraghavan <notifications@github.com> writes:

>> 
>> Sridhar Boovaraghavan <notifications@github.com> writes:
>> > > > @sridharb1 I've submitted those changes using a new PR. I don't think it's possible to reopen a PR after it's been merged and deleted. I tried and couldn't get it to work, so I opened a new PR. #1128 There's something odd about 1.patch.txt which wants to create a new file `mrwthumb.vcxproj` which already exists. I've replaced the old file. > Don't know and can't explain. All I did was to pull down the 0.27-maintenance branch after you had merged it and compared my build directory with the contents of contrib/msvc and found that this was missing. Regarding access, there is a problem that would make this arrangement not sustainable. If I am the maintainer of this directory for the 0.27-maintenance branch, then I need write access to this folder. If not, we are back to square one of me contributing changes, inclusion of which depends on the whims of someone else. I have had that discussion of why I think this is useful and am not inclined to pursue it again with people who disagree. So, this needs to be clear going forward. If I am the maintainer of this (and I take this responsibility for the 0.27 branch because I haven't invested time into supporting changes that are in 0.28+), I need to be able to push into this hierarchy/folder (don't care about this permission in the other folders, the usual process is fine). If I can't, then, no deal.
>> We have established the pull request workflow to ensure that only non-broken commits get into the master or maintenance branches. You already got a pass on the "only tested" part (as your project files are not tested on the CI), so please consider sticking to the pull request workflow as unfortunately github does not provide per folder ACLs. But a human reviewer can easily see and ACK your PR.
>
> That's fair. However, let me propose the following:
>
> 1. link contrib/msvc to my github repository. That is, fork sridharb1/exiv2-x86_x64 to contrib/msvc in the exiv2 repository as a submodule.
> 2. This way, my changes in my repository could be pulled (I could initiate a pull request or release admins can, when they are getting close to a RC1).
> 3. This preserves the control that current administrators have over the exiv2 repository while delinking it from rights that I might need to properly maintain it because I would maintain my own repository.
>
> Is this possible?

As far as I know not easily. Submodules would be an option, but these
don't update automatically and I personally don't find submodules a good
solution for and would prefer not to start using them as they turn into
hard code dependencies and resulting headaches later on.

Would it be highly inconvenient for you on every update that you make to
your fork to just submit the changes in parallel to exiv2/exiv2 too? If
they only touch your msvc files, then I or probably Robin way earlier
will just approve and merge that.

**Comment by clanmills** [comment]
I don't know anything about submodules.  Dan's correct to point out that Exiv2 repos is no longer self contained.  And while contrib/msvc is a leaf of the project today, that might not always be true.  Let's not consider submodules.

Here's another approach.  I will manually test contrib/msvc before every release (RC1..GM) and if there's something wrong, we'll have to cooperate to remedy the situation.  That's how the top-level `msvc/` directory was maintained for about 10 years.

I'll test contrib/msvc today and give you feedback.

**Comment by sridharb1** [comment]
repository. That is, fork sridharb1/exiv2-x86_x64 to contrib/msvc in the exiv2 repository as a submodule. 2. This way, my changes in my repository could be pulled (I could initiate a pull request or release admins can, when they are getting close to a RC1). 3. This preserves the control that current administrators have over the exiv2 repository while delinking it from rights that I might need to properly maintain it because I would maintain my own repository. Is this possible?
> As far as I know not easily. Submodules would be an option, but these don't update automatically and I personally don't find submodules a good solution for and would prefer not to start using them as they turn into hard code dependencies and resulting headaches later on. Would it be highly inconvenient for you on every update that you make to your fork to just submit the changes in parallel to exiv2/exiv2 too? If they only touch your msvc files, then I or probably Robin way earlier will just approve and merge that.
[quoted text above]

You want to avoid submodules, fine. But, please, if it matters, forget about the word submodule and just consider what I have requested. Which is a "manual" link between the repository sridharb1/exiv2-x86_x64 and contrib/msvc. The link should be robust enough that it could be updated on demand without major hassles, ideally through a pull mechanism. The idea I had was a fork between the two which could be updated on demand.

The workflow at the moment looks like this:
1. A new PR is opened to "update" the msvc files.
2. An examination of the difference between my repository and the current contrib/msvc is done and a patch file is generated between the two.
3. This is attached to the PR.
4. One of the maintainers then applies the patch to the contrib/msvc folder and then pushes it to the correct branch after the necessary verification.

This is highly manual and I think with the current git technology, it should be able to be better organized with less manual intervention.

Could you please let me know how this could be done in a better manner?

**Comment by sridharb1** [comment]
> I don't know anything about submodules. Dan's correct to point out that Exiv2 repos is no longer self contained. And while contrib/msvc is a leaf of the project today, that might not always be true. Let's not consider submodules.
> 
> Here's another approach. I will manually test contrib/msvc before every release (RC1..GM) and if there's something wrong, we'll have to cooperate to remedy the situation. That's how the top-level `msvc/` directory was maintained for about 10 years.
> 
> I'll test contrib/msvc today and give you feedback.

Thanks Robin. This is not to sound morbid, but we may not have you tomorrow and I am investigating a way to get it done in a smoother manner.

BTW, as I mentioned, there are a lot of moving parts in compiling exiv2 for windows (I have documented them all in my repository) and it will take a little bit of time to get it done. But, once you set it up, then updating to new versions of dependencies and exiv2 should be a (fairly painless) breeze.

**Comment by clanmills** [comment]
It's not morbid.  Any of us could be gone in the next few months.

Let me see what's going on in `contrib/msvc` and we'll take it from there.  I have a great deal of experience with Visual Studio.  I retired from Adobe 6 years ago **today** and hardly used Visual Studio since.

A bigger challenge for `contrib/msvc` is Exiv2 v0.28 as I agreed with the team that I will not participate in the Exiv2 v0.28 project.  @tbeu likes Visual Studio and has maintained his own sln+vcxproj files for about 10 years.  Perhaps @tbeu will take the challenge of `contrib/msvc` in v0.28.

For now, let's keep it simple.  I'll give you feedback on `msvc/contrib` later today.

**Comment by clanmills** [comment]
I've started to look at `contrib/msvc` and first order of business is to install *VS 2019 Community Edition* on my machine.

I see you have all those references to other projects to build dependencies (zlib, expat, openssl etc).  It's not sure where those go in the file system.  Parallel to exiv2, like this?
```
contrib/msvc/build
contrib/msvc/expat/
```
I'm confident that we'll get this up and working quite quickly.  When I know exactly what you have, we can discuss maintenance.

**Comment by sridharb1** [comment]
No, please put those other projects at the same level as exiv2. That is, E:\Projects\exiv2\build and E:\Projects\libexpat and E:\Projects\openssl etc. My project/solution files usually go into the build folder underneath this top level folder, for e.g., E:\Projects\openssl\build.


Sent from Yahoo Mail for iPhone


On Tuesday, March 31, 2020, 21:22, Robin Mills <notifications@github.com> wrote:




I've started to look at contrib/msvc and first order of business is to install VS 2019 Community Edition on my machine.

I see you have all those references to other projects to build dependencies (zlib, expat, openssl etc). It's not sure where those go in the file system. Parallel to exiv2, like this?
contrib/msvc/build
contrib/msvc/expat/

I'm confident that we'll get this up and working quite quickly. When I know exactly what you have, we can discuss maintenance.

**Comment by clanmills** [comment]
I've installed 2019, updated 2017, installed the tools, updated cmake, had a good night's sleep, fought some mysteries everywhere with success.  I can build exiv2 on 2017 and 2019 on the build server.  I can build/debug both using my script and manually in cmd.exe.  Very pleased.

The folder `contrib\msvc\build` isn't in the source package because the name `build` is in .gitignore.  The easy fix is to rename it `contrib\msvc\exiv2`.  When I open `contrib\msvc\exiv2\exiv2.sln` in VS2019, he's very unhappy.  I can see the structure of the solution, however not a single project loads.  It complains about some COM object returning E_FAIL.

It's going to be a bitter fight to get this working.  I am 100% confident of victory.  However, I don't think this is a priority for v0.27.3.  I think it's more important for me to:
1) Continue to port fixes from master into 0.27-maintenance
2) Work on ISOBMFF (CR3, HEIF, AVIF)

Can you give me your opinion about renaming `build` to `exiv2`.  If you're OK with that, I'll submit a PR later today.

**Comment by clanmills** [comment]
Right.  I've fixed that.  I moved `<exiv2>\contrib\msvc\build` to `<exiv2>\msvc` and open the solution `<exiv2>\msvc\exiv2.sln` and it's fine.  The compiler fails build because I'll need to install the bundles for expat, libz etc... parallel to `<exiv2>`.

We'll have to polish `contrib/msvc/README.md` to explain how to work with this.  It would be useful to have a script `contrib\msvc\setup.bat` to do the git clones (or whatever) to get the dependencies into the correct location.  And that script can recursively copy `contrib\msvc\build` into `<exiv2>\msvc`.   We must rename `contrib\msvc\build` to be anything except _**build**_.  Maybe it should be `contrib\msvc\msvc`. 

I've seen enough to know that I can get this to work.  I'm going to go back to porting fixes from `master` to `0.27-maintenance` and then work on ISOBMFF.  I'm confident about the schedule for RC1 on 2020-04-30.

**Comment by sridharb1** [comment]
> 
> 
> Right. I've fixed that. I moved `<exiv2>\contrib\msvc\build` to `<exiv2>\msvc` and open the solution `<exiv2>\msvc\exiv2.sln` and it's fine. The compiler fails build because I'll need to install the bundles for expat, libz etc... parallel to `<exiv2>`.
> 

Good. The solution and the project files are all based on relative paths. It is currently setup in such a way that exiv2.sln is found in exiv2\build (or msvc, doesn't matter) and openssl/expat are all in the same level as exiv2.

The folder named build is possibly not referred to in the files as well. So, you can change its name.

If you would like to change the hierarchy so that these files live at contrib/msvc/exiv2 in the exiv2 folder, that can be done. I can also change the folder name build to exiv2, update my repository with the instructions. However, I am in the middle of a disk crash/data retrieval situation that will possibly put this out for another day at least.

Regarding the batch script to git clone etc, I wouldn't do that. It can be done, no doubt, but my instructions were all manual and needed to be done one-time. If you are thinking of including this in a build script that would fetch the sources, compile etc., yes, it would need such a script and it can be done. However, I would also ensure that the dependencies are fetched by tags (say, we know openssl1.1.0d works, then don't blindly fetch the latest etc.) to guarantee a successful build.

However, if we are thinking of a batch script to ease these manual actions of the user, then maybe, it could be considered. My feeling is that, you work through it once, get a successful build with the instructions that I have provided (in the my exiv2 project as well as the dependencies) and let's take a call later.

I believe things should be smooth now that you have the hang of how I have organized it.

**Comment by clanmills** [comment]
We can put all the dependencies into SVN.  The user can pull them in with a single command, something like:
```
svn export svn://dev.exiv2.org/svn/team/contrib/msvc/support
```
Open `contrib/msvc/exiv2/exiv2.sln` in Visual Studio 2019 and build!

If the user wants to change the versions of libexpat/zlib/whatever that's their choice.  However, we can be very confident that this builds first time, every time!  Everything is in `contrib/msvc`.  Simple.

```
contrib/msvc/exiv2              Your code
contrib/msvc/support/libexpat   Pulled in from SVN
contrib/msvc/support/zlib       ....
....
```

**Comment by sridharb1** [comment]
Philosophically, this is getting closer to the conan approach of fetching dependencies and using them. For such people, the "standard" build process will suffice. For people, like me, who prefer to have greater control over these things and be able to move parts around independently, my approach of libraries that are not tightly coupled works.

I prefer to not check in any dependencies. Your effort is a "third-person" validating that things work in a "foreign" setup thereby validating the instructions and possibly identifying things that I have missed documenting, which is important.

**Comment by clanmills** [comment]
OK.  I'm going to rename `contrib\msvc\build` as `contrib\msvc2019\solution` as the only file in that directory that's essential is `contrib\msvc2019\solution\exiv2.sln`.

Please change the relative paths in the solution and project files so that the user can open it in VS and the code is there.

README.md has to make it easier to understand where to obtain the dependencies and exactly where they should be located in the file system.

Another thought I had about  the `support` directory is that you could use a junction to "point" to your tree of dependencies.  Other users could "pull" the `support` directory from svn.  Simple. 
```
junction support ..\..\..\..   # or something like that
```

I'll update your stuff and it'll be in Exiv2 v0.27.3.1 (= RC1)  The GM release will be v0.27.3.

I highly recommend speaking to Thomas and getting his input on this.  @tbeu could you work with @sridharb1 on this please.

**Pull request merged**
